pkg/randchar/generator_test.go

```diff
@@ -0,0 +1,430 @@
+		t.Run(name, func(t *testing.T) {
+			options := make([]Option, len(tc.minima))
+			for i, min := range tc.minima {
+				options[i] = Min(min.count, min.charset)
+			}
+
+			rand, err := NewRand(tc.base, options...)
+			assert.Equal(t, err, tc.expectedInitError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			actual, err := rand.Generate(tc.n)
+			assert.Equal(t, err, tc.expectedGenerateError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			for _, min := range tc.minima {
+				count := countFromSet(actual, min.charset)
+				if count < min.count {
+					t.Errorf("unexpected count for %v: %d (actual) < %d (expected minimum)", min.charset, count, min.count)
+				}
+			}
+		})
+	}
+}
+
+func TestMultipleMin(t *testing.T) {
+	cases := map[string]struct {
+		base                  Charset
+		minima                []minimum
+		expectedInitError     error
+		n                     int
+		expectedGenerateError error
+	}{}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			options := make([]Option, len(tc.minima))
+			for i, min := range tc.minima {
+				options[i] = Min(min.count, min.charset)
+			}
+
+			rand, err := NewRand(tc.base, options...)
+			assert.Equal(t, err, tc.expectedInitError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			actual, err := rand.Generate(tc.n)
+			assert.Equal(t, err, tc.expectedGenerateError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			for _, min := range tc.minima {
+				count := countFromSet(actual, min.charset)
+				if count < min.count {
+					t.Errorf("unexpected count for %v: %d (actual) < %d (expected minimum)", min.charset, count, min.count)
+				}
+			}
+		})
+	}
+
+}
+
+func TestNewCharset(t *testing.T) {
+	cases := map[string]struct {
+		in       string
+		expected string
+	}{
+		"empty": {
+			in:       "",
+			expected: "",
+		},
+		"duplicates": {
+			in:       "aa",
+			expected: "a",
+		},
+		"multiple": {
+			in:       "abc",
+			expected: "abc",
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			actual := NewCharset(tc.in)
+
+			if !actual.Equal(Charset(tc.expected)) {
+				t.Errorf("unexpected result: %v (actual) != %v (expected)", actual, tc.expected)
+			}
+		})
+	}
+}
+func TestAdd(t *testing.T) {
+	cases := map[string]struct {
+		a        string
+		b        string
+		expected string
+	}{
+		"equal": {
+			a:        "abc",
+			b:        "abc",
+			expected: "abc",
+		},
+		"subset": {
+			a:        "ab",
+			b:        "abc",
+			expected: "abc",
+		},
+		"superset": {
+			a:        "abc",
+			b:        "ab",
+			expected: "abc",
+		},
+		"empty A": {
+			a:        "",
+			b:        "abc",
+			expected: "abc",
+		},
+		"empty B": {
+			a:        "abc",
+			b:        "",
+			expected: "abc",
+		},
+		"both empty": {
+			a:        "",
+			b:        "",
+			expected: "",
+		},
```

Comment:
How about a testcase for adding sets that are both not a subset of the other for example: "ab", "bc" => "abc"?
In this case neither of the sets is equal to the expected result of the sum of the sets.